DependencyInjection/CmfResourceExtension.php

```diff
@@ -30,125 +44,70 @@ public function load(array $configs, ContainerBuilder $container)
                 ));
             }
 
-            $definition = $this->{$methods[$repository['type']]}($repository['options'], $alias);
-            $definition->addTag('cmf_resource.repository', ['type' => $repository['type'], 'alias' => $alias]);
-            $container->setDefinition('cmf_resource.repository.'.$alias, $definition);
-        }
+            $factory = $this->repositoryFactories[$type];
 
-        $container->setAlias('cmf_resource.registry', 'cmf_resource.registry.container');
-    }
-
-    private function createDoctrinePhpcrOdmRepository(array $options, $alias)
-    {
-        if (!isset($options['basepath'])) {
-            $options['basepath'] = null;
-        }
-
-        $definition = new Definition('Symfony\Cmf\Component\Resource\Repository\PhpcrOdmRepository');
-        $definition->addArgument(new Reference('doctrine_phpcr'));
-        $definition->addArgument($options['basepath']);
-
-        unset($options['basepath']);
-
-        $this->validateRemainingOptions($options, ['basepath'], $alias);
-
-        return $definition;
-    }
-
-    private function createDoctrinePhpcrRepository(array $options, $alias)
-    {
-        if (!isset($options['basepath'])) {
-            $options['basepath'] = null;
-        }
+            $optionsResolver = new OptionsResolver();
+            $factory->configure($optionsResolver);
 
-        $definition = new Definition('Symfony\Cmf\Component\Resource\Repository\PhpcrRepository');
-        $definition->addArgument(new Reference('doctrine_phpcr.session'));
-        $definition->addArgument($options['basepath']);
-
-        unset($options['basepath']);
-
-        $this->validateRemainingOptions($options, ['basepath'], $alias);
-
-        return $definition;
-    }
-
-    private function createCompositeRepository(array $options, $alias)
-    {
-        if (!isset($options['mounts'])) {
-            throw new InvalidConfigurationException('The composite repository type requires a "mounts" option to be set.');
-        }
-
-        $definition = new Definition('Puli\Repository\CompositeRepository');
-
-        foreach ($options['mounts'] as $mount) {
-            if (!isset($mount['mountpoint']) || !isset($mount['repository'])) {
-                throw new InvalidConfigurationException('The "mounts" option of the composite repository type requires a "mountpoint" and "repository" options to be set.');
+            try {
+                $config = $optionsResolver->resolve($config['options']);
+            } catch (\Exception $e) {
+                throw new InvalidArgumentException(sprintf(
+                    'Invalid configuration for repository "%s"',
+                    $repositoryName
+                ), null, $e);
             }
 
-            $definition->addMethodCall('mount', [$mount['mountpoint'], $mount['repository']]);
-        }
-
-        unset($options['mounts']);
+            $serviceId = self::getRepositoryServiceId($repositoryName);
```

Comment:
this is highly based on conventions. Why not make this part of the Factory (i.e. passing the container builder or by introducing a `getRepositoryServiceId` method?).